### PR TITLE
Rbsfxirjtp 162 b calculator service food item price optimization logic

### PR DESCRIPTION
**Pull request by radevdetelin**

### CODE REVIEW

File: calculator/src/main/java/com/market/banica/calculator/dto/ProductPriceComponentsSet.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+    @EqualsAndHashCode.Exclude
+    private final int productId = count.incrementAndGet();
+
+    private String productName;
+
+    private BigDecimal price;
+
+    private Pair<Long,Long> reservedQuantityRange_Start_End = new Pair<>(0L,0L);
+
```
Comment (valto2) [inline calculator/src/main/java/com/market/banica/calculator/dto/ProductPriceComponentsSet.java:28]: Wouldn't be better without underscores in the name?

File: calculator/src/main/java/com/market/banica/calculator/model/Pair.java [hunk 2]
```diff
@@ -0,0 +1,11 @@
+package com.market.banica.calculator.model;
+
+import lombok.AllArgsConstructor;
+import lombok.Data;
+
+@Data
+@AllArgsConstructor
+public class Pair  <T,K>{
+    private T fst;
+    private K snd;
```
Comment (valto2) [inline calculator/src/main/java/com/market/banica/calculator/model/Pair.java:10]: Names could be more descriptive.

File: calculator/src/main/java/com/market/banica/calculator/service/CalculatorServiceImpl.java [hunk 3]
```diff
@@ -1,56 +1,823 @@
...
+        Set<ProductPriceComponentsSet> simpleProductPriceVariantsSet = calculatePossiblePricesForProductFromProductSpecifications(
+                productDtoMap, productSpecificationMap, orderedProduct,
+                orderedProduct, productPriceComponentsSetByProductIdMap);
+
+        List<List<ProductPriceComponentsSet>> compositeProductPriceVariantSets = result.get(orderedProduct.getItemName());
+
+        List<ProductPriceComponentsSet> compositeIngredientPriceVariants = compositeProductPriceVariantSets
+                .stream()
+                .flatMap(Collection::stream)
+                .filter(l -> l.getProductName().equals(orderedProduct.getItemName()))
```
Comment (valto2) [inline calculator/src/main/java/com/market/banica/calculator/service/CalculatorServiceImpl.java:385]: " l " could be changed with something more descriptive.

File: calculator/src/main/java/com/market/banica/calculator/service/CalculatorServiceImpl.java [hunk 4]
```diff
@@ -1,56 +1,823 @@
...
+                .collect(Collectors.toList());
+
+        compositeIngredientPriceVariants.forEach(m -> {
+            BigDecimal newPrice = m.getPrice().multiply(BigDecimal.valueOf(quantity));
+            m.setPrice(newPrice);
```
Comment (valto2) [inline calculator/src/main/java/com/market/banica/calculator/service/CalculatorServiceImpl.java:390]: " m " could be changed with something more descriptive.

File: calculator/src/main/java/com/market/banica/calculator/service/ProductServiceImpl.java [hunk 5]
```diff
@@ -197,18 +200,10 @@ private void removeDeletedProductFromAllRecipes(String productName) {
...
-        validateProductExists(productName);
-
-        return productBase.getDatabase().get(productName);
-    }
-
     private void writeProductToDatabase(String newProductName, Product newProduct) {
         LOGGER.debug("In writeProductToDatabase private method");
 
-        announceInterestToOrderBookProductBase(newProductName);
+//        announceInterestToOrderBookProductBase(newProductName);
```
Comment (valto2) [inline calculator/src/main/java/com/market/banica/calculator/service/ProductServiceImpl.java:206]: Probably forgotten to be deleted ?

File: calculator/src/test/java/com/market/banica/calculator/controllerTests/CalculatorControllerTest.java [hunk 6]
```diff
@@ -44,13 +46,15 @@ void getRecipe() throws Exception {
...
+        List<ProductDto> productDtoList = new ArrayList<>();
+        ProductDto dummyRecipe = new ProductDto();
+        dummyRecipe.setIngredients(new HashMap<>());
         dummyRecipe.setItemName("baklava");
         dummyRecipe.setTotalPrice(BigDecimal.valueOf(10));
+        productDtoList.add(dummyRecipe);
 
 
-        given(service.getRecipe(clientId, product, 100)).willReturn(dummyRecipe);
+        given(service.getRecipe(clientId, product, 100)).willReturn(productDtoList);
```
Comment (valto2) [inline calculator/src/test/java/com/market/banica/calculator/controllerTests/CalculatorControllerTest.java:57]: "100" could be extracted in constant.

File: calculator/src/test/java/com/market/banica/calculator/service/ProductServiceImplTest.java [hunk 7]
```diff
@@ -1,326 +1,327 @@
...
-    }
-
-    private Map<String, Product> getDatabaseWithIngredients() {
-        Map<String, Product> ingredients = new ConcurrentHashMap<>();
-        ingredients.put(PUMPKIN, createProduct(PUMPKIN));
-        return ingredients;
-    }
-}
\ No newline at end of file
+//package com.market.banica.calculator.service;
```
Comment (valto2) [inline calculator/src/test/java/com/market/banica/calculator/service/ProductServiceImplTest.java:1]: Is there a reason for the tests to be commented ?

File: calculator/src/main/java/com/market/banica/calculator/dto/ProductPriceComponentsSet.java (hunk 1 above)
Comment (radevdetelin) [reply calculator/src/main/java/com/market/banica/calculator/dto/ProductPriceComponentsSet.java]: The point was to emphasise what the Pair components are. It was applied for code readability, I agree it is not consistent with the team code style. Fixed

File: calculator/src/main/java/com/market/banica/calculator/model/Pair.java (hunk 2 above)
Comment (radevdetelin) [reply calculator/src/main/java/com/market/banica/calculator/model/Pair.java:10]: As it is a generic class with several use cases in the code, field names have to be generic too. I applied Sun's Pair class convention fst for the first and snd for the second. Do you mean to replace it with "first", respectively "second"?

Comment (valto2) [reply calculator/src/main/java/com/market/banica/calculator/model/Pair.java:10]: If that's the reason , then it is ok to leave it as it is.

File: calculator/src/main/java/com/market/banica/calculator/service/CalculatorServiceImpl.java (hunk 3 above)
Comment (radevdetelin) [reply calculator/src/main/java/com/market/banica/calculator/service/CalculatorServiceImpl.java:385]: With respect to the naming convention of the team Oracle Code  Convention says:
"Variable names should be short yet meaningful. The choice of a variable name should be mnemonic- that is, designed to indicate to the casual observer the intent of its use. **One-character variable names should be avoided except for temporary "throwaway" variables.** Common names for temporary variables are i, j, k, m, and n for integers; c, d, and e for characters."

File: calculator/src/main/java/com/market/banica/calculator/service/CalculatorServiceImpl.java (hunk 4 above)
Comment (radevdetelin) [reply calculator/src/main/java/com/market/banica/calculator/service/CalculatorServiceImpl.java:390]: With respect to the naming convention of the team Oracle Code Convention says:
"Variable names should be short yet meaningful. The choice of a variable name should be mnemonic- that is, designed to indicate to the casual observer the intent of its use. One-character variable names should be avoided except for temporary "throwaway" variables. Common names for temporary variables are i, j, k, m, and n for integers; c, d, and e for characters."

File: calculator/src/main/java/com/market/banica/calculator/service/ProductServiceImpl.java (hunk 5 above)
Comment (radevdetelin) [reply calculator/src/main/java/com/market/banica/calculator/service/ProductServiceImpl.java:206]: It is a part of the testing scenario, as I am still doing intensive testing, it is to be fixed before the actual merge, as it is only redirecting data from actual instances to test instances, probably tomorrow morning

File: calculator/src/test/java/com/market/banica/calculator/controllerTests/CalculatorControllerTest.java (hunk 6 above)
Comment (radevdetelin) [reply calculator/src/test/java/com/market/banica/calculator/controllerTests/CalculatorControllerTest.java:57]: It is not part of my code, but from the merge commit with main

File: calculator/src/test/java/com/market/banica/calculator/service/ProductServiceImplTest.java (hunk 7 above)
Comment (radevdetelin) [reply calculator/src/test/java/com/market/banica/calculator/service/ProductServiceImplTest.java]: Fixed

File: calculator/src/main/java/com/market/banica/calculator/service/CalculatorServiceImpl.java (hunk 3 above)
Comment (gaantonova) [reply calculator/src/main/java/com/market/banica/calculator/service/CalculatorServiceImpl.java]: in this case, the variable is an object, please change it to a more meaningful name

File: calculator/src/main/java/com/market/banica/calculator/service/CalculatorServiceImpl.java (hunk 4 above)
Comment (gaantonova) [reply calculator/src/main/java/com/market/banica/calculator/service/CalculatorServiceImpl.java]: same here, those rules are meant for example when you have a for loop

File: calculator/src/main/java/com/market/banica/calculator/service/ProductServiceImpl.java (hunk 5 above)
Comment (gaantonova) [reply calculator/src/main/java/com/market/banica/calculator/service/ProductServiceImpl.java]: If this is not going into the main branch, please delete it from the PR and leave it locally for you to test

File: calculator/src/test/java/com/market/banica/calculator/controllerTests/CalculatorControllerTest.java (hunk 6 above)
Comment (gaantonova) [reply calculator/src/test/java/com/market/banica/calculator/controllerTests/CalculatorControllerTest.java:57]: It doesn't matter if you didn't write it initially, our goal is to improve the code quality whenever possible, so please change it

File: calculator/src/main/java/com/market/banica/calculator/model/Pair.java (hunk 2 above)
Comment (gaantonova) [reply calculator/src/main/java/com/market/banica/calculator/model/Pair.java:10]: Why not use the already available Pair class located in javafx.util ?

Comment (zaraandreevaa) [reply calculator/src/main/java/com/market/banica/calculator/model/Pair.java:10]: the name of the variables should be descriptive, at least a word, first and second sound better
also, I can see that the class is not formatted

File: calculator/src/test/java/com/market/banica/calculator/controllerTests/CalculatorControllerTest.java [hunk 8]
```diff
@@ -44,13 +46,15 @@ void getRecipe() throws Exception {
         String product = "baklava";
         int quantity = 100;
 
-        RecipeDTO dummyRecipe = new RecipeDTO();
-        dummyRecipe.setIngredients(new HashSet<>());
+        List<ProductDto> productDtoList = new ArrayList<>();
```
Comment (zaraandreevaa) [inline calculator/src/test/java/com/market/banica/calculator/controllerTests/CalculatorControllerTest.java:49]: population of the list could be extracted in a separate method and all the strings should be in a constant

File: calculator/src/main/java/com/market/banica/calculator/controller/CalculatorController.java [hunk 9]
```diff
@@ -24,19 +26,16 @@
...
-        //replace new RecipeDTO with service call when service is completed
-        RecipeDTO recipeDTO = service.getRecipe(clientId, itemName, quantity);
+    public List<ProductDto> getRecipe(@PathVariable("clientId") @NotBlank String clientId,
+                                                @PathVariable("itemName") @NotBlank String itemName,
+                                                @PathVariable("quantity") @Min(1) long quantity) {
+        LOGGER.info("GET /calculator called");
 
 
-        return ResponseEntity.ok().body(recipeDTO);
+        return service.getRecipe(clientId,itemName, quantity);
```
Comment (zaraandreevaa) [inline calculator/src/main/java/com/market/banica/calculator/controller/CalculatorController.java:39]: the code is not formatted

File: calculator/src/main/java/com/market/banica/calculator/dto/ProductDto.java [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+
+    private BigDecimal totalPrice = BigDecimal.ZERO;
+
+    private List<ProductSpecification> productSpecifications = new ArrayList<>();
+
+    private Map<String, Long> ingredients = new HashMap<>();
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
```
Comment (zaraandreevaa) [inline calculator/src/main/java/com/market/banica/calculator/dto/ProductDto.java:25]: curly braces are missing

File: calculator/src/main/java/com/market/banica/calculator/dto/ProductDto.java [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+        if (!(o instanceof ProductDto)) return false;
```
Comment (zaraandreevaa) [inline calculator/src/main/java/com/market/banica/calculator/dto/ProductDto.java:26]: same here, curly braces

File: calculator/src/main/java/com/market/banica/calculator/dto/ProductPriceComponentsSet.java [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+    private static final AtomicInteger count = new AtomicInteger(0);
+
+    @EqualsAndHashCode.Exclude
+    private final int productId = count.incrementAndGet();
+
+    private String productName;
+
+    private BigDecimal price;
+
+    private Pair<Long,Long> reservedQuantityRangeStartEnd = new Pair<>(0L,0L);
```
Comment (zaraandreevaa) [inline calculator/src/main/java/com/market/banica/calculator/dto/ProductPriceComponentsSet.java:27]: not formatted code

File: calculator/src/main/java/com/market/banica/calculator/model/Product.java [hunk 13]
```diff
@@ -6,14 +6,28 @@
...
     private String productName;
 
     private UnitOfMeasure unitOfMeasure;
 
-    private Map<String, Integer> ingredients = new HashMap<>();
+    private Map<String,Long> ingredients = new HashMap<>();
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
```
Comment (zaraandreevaa) [inline calculator/src/main/java/com/market/banica/calculator/model/Product.java:23]: curly braces

File: calculator/src/main/java/com/market/banica/calculator/service/ProductServiceImpl.java [hunk 14]
```diff
@@ -168,27 +167,29 @@ public void deleteProductFromParentIngredients(String parentProductName, String
...
         return result;
     }
 
-    //TODO to be implemented once expectations are clear
     @Override
-    public void getAllProductsAsListProduct() {
+    public Product getProductFromDatabase(String productName) {
+        LOGGER.debug("In getProductFromDatabase method");
+
+        validateProductExists(productName);
```
Comment (zaraandreevaa) [inline calculator/src/main/java/com/market/banica/calculator/service/ProductServiceImpl.java:190]: keyword is missing for calling the method

File: calculator/src/test/java/com/market/banica/calculator/componentTests/CalculatorComponentIT.java [hunk 15]
```diff
@@ -274,10 +280,20 @@ public void cleanUp() {
     }
 
     @NotNull
-    private RecipeDTO createRecipeDTO() {
-        RecipeDTO response = new RecipeDTO();
+    private ProductDto createProductDTO() {
+        ProductDto response = new ProductDto();
+        ProductSpecification productSpecification = new ProductSpecification();
```
Comment (zaraandreevaa) [inline calculator/src/test/java/com/market/banica/calculator/componentTests/CalculatorComponentIT.java:285]: population of ProductSpecification is not part of createProductDTO, it should be in a separate method

**Review by zaraandreevaa — COMMENTED** [review]
almost all the classes are not formatted, please fix all of them to not write the same comment per each class

File: calculator/src/main/java/com/market/banica/calculator/model/Pair.java (hunk 2 above)
Comment (radevdetelin) [reply calculator/src/main/java/com/market/banica/calculator/model/Pair.java:10]: @gaantonova It is no longer part of the Java JDK,  you need to download the JavaFX SDK to your machine 
@zaraandreevaa fixed

File: calculator/src/test/java/com/market/banica/calculator/controllerTests/CalculatorControllerTest.java (hunk 6 above)
Comment (radevdetelin) [reply calculator/src/test/java/com/market/banica/calculator/controllerTests/CalculatorControllerTest.java:57]: Fixed

File: calculator/src/main/java/com/market/banica/calculator/service/ProductServiceImpl.java (hunk 5 above)
Comment (radevdetelin) [reply calculator/src/main/java/com/market/banica/calculator/service/ProductServiceImpl.java]: Fixed

File: calculator/src/test/java/com/market/banica/calculator/controllerTests/CalculatorControllerTest.java (hunk 8 above)
Comment (radevdetelin) [reply calculator/src/test/java/com/market/banica/calculator/controllerTests/CalculatorControllerTest.java:49]: Fixed

File: calculator/src/main/java/com/market/banica/calculator/controller/CalculatorController.java (hunk 9 above)
Comment (radevdetelin) [reply calculator/src/main/java/com/market/banica/calculator/controller/CalculatorController.java:39]: Fixed

File: calculator/src/main/java/com/market/banica/calculator/dto/ProductDto.java (hunk 10 above)
Comment (radevdetelin) [reply calculator/src/main/java/com/market/banica/calculator/dto/ProductDto.java:25]: Fixed

File: calculator/src/main/java/com/market/banica/calculator/dto/ProductDto.java (hunk 11 above)
Comment (radevdetelin) [reply calculator/src/main/java/com/market/banica/calculator/dto/ProductDto.java:26]: Fixed

File: calculator/src/main/java/com/market/banica/calculator/dto/ProductPriceComponentsSet.java (hunk 12 above)
Comment (radevdetelin) [reply calculator/src/main/java/com/market/banica/calculator/dto/ProductPriceComponentsSet.java:27]: Fixed

File: calculator/src/main/java/com/market/banica/calculator/model/Product.java (hunk 13 above)
Comment (radevdetelin) [reply calculator/src/main/java/com/market/banica/calculator/model/Product.java:23]: Fixed

File: calculator/src/main/java/com/market/banica/calculator/service/ProductServiceImpl.java (hunk 14 above)
Comment (radevdetelin) [reply calculator/src/main/java/com/market/banica/calculator/service/ProductServiceImpl.java:190]: Fixed

File: calculator/src/test/java/com/market/banica/calculator/componentTests/CalculatorComponentIT.java (hunk 15 above)
Comment (radevdetelin) [reply calculator/src/test/java/com/market/banica/calculator/componentTests/CalculatorComponentIT.java:285]: Fixed

File: calculator/src/main/java/com/market/banica/calculator/service/CalculatorServiceImpl.java (hunk 4 above)
Comment (radevdetelin) [reply calculator/src/main/java/com/market/banica/calculator/service/CalculatorServiceImpl.java]: Fixed

File: calculator/src/main/java/com/market/banica/calculator/service/CalculatorServiceImpl.java (hunk 3 above)
Comment (radevdetelin) [reply calculator/src/main/java/com/market/banica/calculator/service/CalculatorServiceImpl.java]: Fixed

**Review by valto2 — APPROVED** [review]

File: calculator/pom.xml [hunk 16]
```diff
@@ -149,6 +149,13 @@
             <groupId>com.market.banica</groupId>
             <artifactId>common</artifactId>
         </dependency>
+
+        <dependency>
+            <groupId>org.apache.maven.plugins</groupId>
+            <artifactId>maven-surefire-plugin</artifactId>
+            <version>3.0.0-M4</version>
```
Comment (zaraandreevaa) [inline calculator/pom.xml:156]: please extract the version in properties

File: calculator/src/main/java/com/market/banica/calculator/service/ProductServiceImpl.java [hunk 17]
```diff
@@ -313,36 +306,47 @@ private String getProductName(List<Product> products) {
...
+        Queue<Product> tempContainer = new ArrayDeque<>();
+        tempContainer.add(parentProduct);
 
         while (!tempContainer.isEmpty()) {
 
-            Product tempProduct = tempContainer.remove();
+            Product tempParentProduct = tempContainer.remove();
 
-            if (!tempProduct.getIngredients().isEmpty()) {
+            if (!tempParentProduct.getIngredients().isEmpty()) {
```
Comment (zaraandreevaa) [inline calculator/src/main/java/com/market/banica/calculator/service/ProductServiceImpl.java:320]: a null check on tempParentProduct.getIngredients can be added as well

File: calculator/src/main/java/com/market/banica/calculator/service/ProductServiceImpl.java [hunk 18]
```diff
@@ -313,36 +306,47 @@ private String getProductName(List<Product> products) {
...
+                        }});
+                    }
+                });
             }
         }
     }
 
-    private Queue<Product> convertListOfProductNamesInQueueOfProducts(Product recipe) {
-        LOGGER.debug("In convertListOfProductNamesInQueueOfProducts private method");
+    private Collection<Product> convertProductIngredientsNamesToCollectionOfProducts(Product parent) {
```
Comment (zaraandreevaa) [inline calculator/src/main/java/com/market/banica/calculator/service/ProductServiceImpl.java:343]: productParent sounds better as a variable name

File: calculator/src/test/java/com/market/banica/calculator/componentTests/CalculatorComponentIT.java [hunk 19]
```diff
@@ -274,15 +280,36 @@ public void cleanUp() {
     }
 
     @NotNull
-    private RecipeDTO createRecipeDTO() {
-        RecipeDTO response = new RecipeDTO();
+    private ProductDto createProductDTO() {
+        ProductDto response = new ProductDto();
+
+        ProductSpecification productSpecification = createProductSpecification();
```
Comment (zaraandreevaa) [inline calculator/src/test/java/com/market/banica/calculator/componentTests/CalculatorComponentIT.java:286]: you do not need to create an object, you can directly pass the method where it is used

File: calculator/src/test/java/com/market/banica/calculator/componentTests/CalculatorComponentIT.java [hunk 20]
```diff
@@ -274,15 +280,36 @@ public void cleanUp() {
...
+
         return response;
     }
 
+    @NotNull
+    private ProductSpecification createProductSpecification() {
+
+        ProductSpecification productSpecification = new ProductSpecification();
+
+        productSpecification.setLocation("AMERICA");
```
Comment (zaraandreevaa) [inline calculator/src/test/java/com/market/banica/calculator/componentTests/CalculatorComponentIT.java:302]: please extract in a constant

File: calculator/pom.xml (hunk 16 above)
Comment (radevdetelin) [reply calculator/pom.xml:156]: Fixed

File: calculator/src/main/java/com/market/banica/calculator/service/ProductServiceImpl.java (hunk 17 above)
Comment (radevdetelin) [reply calculator/src/main/java/com/market/banica/calculator/service/ProductServiceImpl.java:320]: Fixed

File: calculator/src/main/java/com/market/banica/calculator/service/ProductServiceImpl.java [hunk 21]
```diff
@@ -313,36 +306,47 @@ private String getProductName(List<Product> products) {
...
-                result.addAll(tempIngredientsQueue);
-
-            } else {
-
-                result.add(tempProduct);
+                tempIngredients.forEach(ingredient -> {
+                    if (productQuantitiesMap.containsKey(ingredient)) {
+                        productQuantitiesMap
+                                .get(ingredient).put(tempParentProduct.getProductName()
+                                , tempParentProduct.getIngredients().get(ingredient.getProductName()));
```
[the remark this answers is not in the document]
Comment (radevdetelin) [reply calculator/src/main/java/com/market/banica/calculator/service/ProductServiceImpl.java:331]: All the classes are formatted, that case is not covered by auto-format. I fixed it manually.

File: calculator/src/main/java/com/market/banica/calculator/service/ProductServiceImpl.java (hunk 18 above)
Comment (radevdetelin) [reply calculator/src/main/java/com/market/banica/calculator/service/ProductServiceImpl.java:343]: Done

File: calculator/src/test/java/com/market/banica/calculator/componentTests/CalculatorComponentIT.java (hunk 19 above)
Comment (radevdetelin) [reply calculator/src/test/java/com/market/banica/calculator/componentTests/CalculatorComponentIT.java:286]: Fixed

File: calculator/src/test/java/com/market/banica/calculator/componentTests/CalculatorComponentIT.java (hunk 20 above)
Comment (radevdetelin) [reply calculator/src/test/java/com/market/banica/calculator/componentTests/CalculatorComponentIT.java:302]: Fixed

**Review by Lilyanov — APPROVED** [review]

**Review by Lilyanov — APPROVED** [review]

**Pull request merged**